### PR TITLE
Windows: Don't close client stdin handle to avoid hang

### LINKED ISSUE #28267: docker run hangs after container exit when started in a new Windows console
@darrenstahlmsft 

docker run on Windows hangs after container exit when started in a new console, from an HCS point of view, the container has exited completely as well.

Repro Steps:
PS C:\container> start docker -ArgumentList 'run -it --isolation hyperv nanoserver cmd'
- or -
C:\Users\vmadmin>start docker run -it --isolation hyperv 824aa583 cmd

I would expect the newly created window to close or exit

Additional Info:
It doesn’t repro if I run it in the same console, i.e. docker run -it --isolation hyperv nanoserver cmd (no start)
It also doesn’t seem matter if I use powershell or plain old cmd.exe, 
Repro's with both --isolation hyperv and --isolation process.

docker version
Client:
 Version:      1.13.0-dev
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   da0ccf8
 Built:        Wed Nov  9 20:37:50 2016
 OS/Arch:      windows/amd64

Server:
 Version:             1.13.0-dev
 API version:         1.25
 Minimum API version: 1.24
 Go version:          go1.7.3
 Git commit:          da0ccf8
 Built:               Wed Nov  9 20:37:50 2016
 OS/Arch:             windows/amd64
 Experimental:        false

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard (VM) <jhoward@microsoft.com>

Fixes https://github.com/docker/docker/issues/28267

@darrenstahlmsft @johnstep PTAL.

TL;DR - Doing the same fix on Windows as the hack on Darwin in the client hijack. It's a different underlying cause (looks like a platform bug on Windows), but as the process is exiting anyway, there's no need to explicitly close the handle. This avoids the client hanging under certain circumstances at the point it would be exiting.

Longer version:

Duplicating the big comment from https://github.com/docker/docker/issues/28267#issuecomment-288237442:

Got this down to `restoreTerminal` not returning in win32 trying to close the processes stdin handle. There's an interesting comment in the code (cli\command\container\hijack.go, restoreTerminal())

```
	// WARNING: DO NOT REMOVE THE OS CHECK !!!
	// For some reason this Close call blocks on darwin..
	// As the client exists right after, simply discard the close
	// until we find a better solution.
```

I'm going to extend the check to exclude Windows as that does fix it. It's almost certainly a different underlying cause to Darwin, but just interesting none-the-less.

Some notes - you only hit this if
a) Using the native console (not the v1/legacy console)
b) You "start" the process in a new window. ie `start docker run --rm -it microsoft/nanoserver cmd /s /c echo foobar` will hang, docker run --rm -it microsoft/nanoserver cmd /s /c echo foobar`  will be fine.

`docker.exe` is stuck in a `CloseHandle` (`ntdll!ZwClose`) in the following stack trace - thread 8 - by the look of it.  I'll follow up with the conhost folks.

```
0:014> ~*k

   0  Id: 1270.874 Suspend: 1 Teb: 00000000`00338000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0008fb58 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
*** WARNING: Unable to verify timestamp for e:\go\src\github.com\docker\docker\bundles\docker.exe
*** ERROR: Module load completed but symbols could not be loaded for e:\go\src\github.com\docker\docker\bundles\docker.exe
01 00000000`0008fb60 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0008fc00 00000000`00e0fc00 docker+0x5d47a
03 00000000`0008fc08 00000000`00000000 docker+0xa0fc00

   1  Id: 1270.1054 Suspend: 1 Teb: 00000000`00340000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`031afc28 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`031afc30 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`031afcd0 000000c0`4202a320 docker+0x5d47a
03 00000000`031afcd8 00000000`00340000 0x000000c0`4202a320
04 00000000`031afce0 00000000`00000000 0x340000

   2  Id: 1270.1884 Suspend: 1 Teb: 00000000`00342000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`031cfdf8 00007ffe`78344310 docker+0x2ba65
01 00000000`031cfe00 00000000`00410e59 KERNEL32!SetEvent
02 00000000`031cfe08 00000000`00e0ecd0 docker+0x10e59
03 00000000`031cfe10 00000000`031cfe48 docker+0xa0ecd0
04 00000000`031cfe18 00000000`004111ec 0x31cfe48
05 00000000`031cfe20 000000c0`4246e400 docker+0x111ec
06 00000000`031cfe28 000000c0`4246e400 0x000000c0`4246e400
07 00000000`031cfe30 00000000`00000001 0x000000c0`4246e400
08 00000000`031cfe38 00000000`00000001 0x1
09 00000000`031cfe40 000000c0`4246e400 0x1
0a 00000000`031cfe48 00000000`031cfe80 0x000000c0`4246e400
0b 00000000`031cfe50 00000000`00433ef7 0x31cfe80
0c 00000000`031cfe58 000000c0`4246e508 docker+0x33ef7
0d 00000000`031cfe60 00000000`00000201 0x000000c0`4246e508
0e 00000000`031cfe68 00000000`031cfea8 0x201
0f 00000000`031cfe70 000000c0`42013500 0x31cfea8
10 00000000`031cfe78 000000c0`4246e400 0x000000c0`42013500
11 00000000`031cfe80 00000000`031cfea8 0x000000c0`4246e400
12 00000000`031cfe88 00000000`0043437e 0x31cfea8
13 00000000`031cfe90 000000c0`42013500 docker+0x3437e
14 00000000`031cfe98 00000001`00000001 0x000000c0`42013500
15 00000000`031cfea0 000000c0`00000001 0x00000001`00000001
16 00000000`031cfea8 00000000`031cfed8 0x000000c0`00000001
17 00000000`031cfeb0 00000000`004316c0 0x31cfed8
18 00000000`031cfeb8 00000000`00e0ed04 docker+0x316c0
19 00000000`031cfec0 000000c0`00000000 docker+0xa0ed04
1a 00000000`031cfec8 000000c0`42014a01 0x000000c0`00000000
1b 00000000`031cfed0 000000c0`4201e340 0x000000c0`42014a01
1c 00000000`031cfed8 00000000`031cff00 0x000000c0`4201e340
1d 00000000`031cfee0 00000000`00453e3f 0x31cff00
1e 00000000`031cfee8 000000c0`423d7520 docker+0x53e3f
1f 00000000`031cfef0 00000000`00000000 0x000000c0`423d7520

   3  Id: 1270.19b8 Suspend: 1 Teb: 00000000`00344000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`031efd98 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`031efda0 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`031efe40 000000c0`4202ab20 docker+0x5d47a
03 00000000`031efe48 000000c0`4203b5d8 0x000000c0`4202ab20
04 00000000`031efe50 00000002`00000000 0x000000c0`4203b5d8
05 00000000`031efe58 00000000`000001d4 0x00000002`00000000
06 00000000`031efe60 00000000`0043b1f6 0x1d4
07 00000000`031efe68 000000c0`420125c4 docker+0x3b1f6
08 00000000`031efe70 00000000`00000001 0x000000c0`420125c4
09 00000000`031efe78 00000001`00000001 0x1
0a 00000000`031efe80 00000000`00411144 0x00000001`00000001
0b 00000000`031efe88 00000000`00e0ecd0 docker+0x11144
0c 00000000`031efe90 00000000`00000001 docker+0xa0ecd0
0d 00000000`031efe98 00000000`00000000 0x1

   4  Id: 1270.13f0 Suspend: 1 Teb: 00000000`00346000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0320fb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`0320fb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0320fc30 000000c0`42036320 docker+0x5d47a
03 00000000`0320fc38 00000000`00000000 0x000000c0`42036320

   5  Id: 1270.18d8 Suspend: 1 Teb: 00000000`00348000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`033cf398 00007ffe`780e8bff ntdll!ZwRemoveIoCompletionEx+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 2971] 
01 00000000`033cf3a0 00000000`0045d47a KERNELBASE!GetQueuedCompletionStatusEx+0xaf [d:\rs1\minkernel\kernelbase\error.c @ 986] 
02 00000000`033cf440 00000000`033cf440 docker+0x5d47a
03 00000000`033cf448 00000000`033cf5e0 0x33cf440
04 00000000`033cf450 00000000`033cf470 0x33cf5e0
05 00000000`033cf458 000000c0`424966f8 0x33cf470
06 00000000`033cf460 00000000`ffffffff 0x000000c0`424966f8
07 00000000`033cf468 00000000`00000000 0xffffffff

   6  Id: 1270.13c0 Suspend: 1 Teb: 00000000`0034a000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0342fdb8 00007ffe`780ac264 ntdll!ZwReadFile+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 227] 
01 00000000`0342fdc0 00000000`0045d47a KERNELBASE!ReadFile+0x74 [d:\rs1\minkernel\kernelbase\filehops.c @ 1397] 
02 00000000`0342fe40 00000000`00000044 docker+0x5d47a
03 00000000`0342fe48 000000c0`423f6368 0x44
04 00000000`0342fe50 00000000`00002710 0x000000c0`423f6368
05 00000000`0342fe58 000000c0`42235968 0x2710
06 00000000`0342fe60 00000000`00000000 0x000000c0`42235968

   7  Id: 1270.14a8 Suspend: 1 Teb: 00000000`0034c000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0344fb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`0344fb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0344fc30 000000c0`423f7320 docker+0x5d47a
03 00000000`0344fc38 00000000`00000000 0x000000c0`423f7320

   8  Id: 1270.1784 Suspend: 1 Teb: 00000000`0034e000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`034bfe08 00007ffe`780adfd2 ntdll!ZwClose+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 299] 
01 00000000`034bfe10 00000000`0045d47a KERNELBASE!CloseHandle+0x62 [d:\rs1\minkernel\kernelbase\handle.c @ 159] 
02 00000000`034bfe40 00000000`00000054 docker+0x5d47a
03 00000000`034bfe48 00000000`034bfee8 0x54
04 00000000`034bfe50 00000000`00000008 0x34bfee8
05 00000000`034bfe58 000000c0`4246a368 0x8
06 00000000`034bfe60 00000000`00000000 0x000000c0`4246a368

   9  Id: 1270.608 Suspend: 1 Teb: 00000000`00350000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`034dfb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`034dfb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`034dfc30 000000c0`4246e320 docker+0x5d47a
03 00000000`034dfc38 00000000`00000000 0x000000c0`4246e320

  10  Id: 1270.1920 Suspend: 1 Teb: 00000000`00352000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0353fb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`0353fb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0353fc30 000000c0`4246e720 docker+0x5d47a
03 00000000`0353fc38 00000000`00000000 0x000000c0`4246e720

  11  Id: 1270.18b4 Suspend: 1 Teb: 00000000`00354000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0355fb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`0355fb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0355fc30 000000c0`42276320 docker+0x5d47a
03 00000000`0355fc38 00000000`00010000 0x000000c0`42276320
04 00000000`0355fc40 00000000`00000000 0x10000

  12  Id: 1270.810 Suspend: 1 Teb: 00000000`00356000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0357fb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`0357fb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0357fc30 000000c0`42036b20 docker+0x5d47a
03 00000000`0357fc38 00000000`00000000 0x000000c0`42036b20

  13  Id: 1270.1b24 Suspend: 1 Teb: 00000000`00358000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0359fb88 00007ffe`780975ff ntdll!ZwWaitForSingleObject+0x14 [d:\rs1.obj.amd64fre\minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 211] 
01 00000000`0359fb90 00000000`0045d47a KERNELBASE!WaitForSingleObjectEx+0x8f [d:\rs1\minkernel\kernelbase\synch.c @ 1328] 
02 00000000`0359fc30 000000c0`423f7720 docker+0x5d47a
03 00000000`0359fc38 00000000`00000000 0x000000c0`423f7720

# 14  Id: 1270.5c0 Suspend: 1 Teb: 00000000`0035a000 Unfrozen
 # Child-SP          RetAddr           Call Site
00 00000000`0363ff28 00007ffe`7ae019ba ntdll!DbgBreakPoint [d:\rs1\minkernel\ntos\rtl\amd64\debugstb.asm @ 51] 
01 00000000`0363ff30 00007ffe`78328364 ntdll!DbgUiRemoteBreakin+0x4a [d:\rs1\minkernel\ntdll\dlluistb.c @ 312] 
02 00000000`0363ff60 00007ffe`7ad970d1 KERNEL32!BaseThreadInitThunk+0x14 [d:\rs1\base\win32\client\thread.c @ 64] 
03 00000000`0363ff90 00000000`00000000 ntdll!RtlUserThreadStart+0x21 [d:\rs1\minkernel\ntdll\rtlstrt.c @ 997] 

```

And for completeness, here's the matching docker stacks

```
(dlv) goroutines
[25 goroutines]
  Goroutine 1 - User: e:/go/src/github.com/docker/docker/cli/command/container/run.go:200 github.com/docker/docker/cli/command/container.runContainer (0x627d45)
  Goroutine 2 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 3 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 17 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 18 - User: C:/Go/src/runtime/sigqueue.go:116 os/signal.signal_recv (0x4419e5)
  Goroutine 20 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 21 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 22 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 23 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 24 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 25 - User: e:/go/src/github.com/docker/docker/pkg/term/term_windows.go:229 github.com/docker/docker/pkg/term.restoreAtInterrupt.func1 (0x4a5474)
* Goroutine 26 - User: C:/Go/src/runtime/syscall_windows.go:163 syscall.Syscall (0x4491e6)
  Goroutine 27 - User: C:/Go/src/runtime/syscall_windows.go:174 syscall.Syscall6 (0x449286)
  Goroutine 28 - User: e:/go/src/github.com/docker/docker/cli/command/container/utils.go:74 github.com/docker/docker/cli/command/container.waitExitOrRemoved.func2.1 (0x636eb4)
  Goroutine 31 - User: C:/Go/src/runtime/netpoll.go:160 net.runtime_pollWait (0x4291f9)
  Goroutine 32 - User: C:/Go/src/net/http/transport.go:1646 net/http.(*persistConn).writeLoop (0x594f81)
  Goroutine 33 - User: C:/Go/src/runtime/asm_amd64.s:576 runtime.asmcgocall (0x45b667)
  Goroutine 35 - User: C:/Go/src/net/http/transport.go:1541 net/http.(*persistConn).readLoop (0x593c14)
  Goroutine 36 - User: C:/Go/src/net/http/transport.go:1646 net/http.(*persistConn).writeLoop (0x594f81)
  Goroutine 37 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 38 - User: C:/Go/src/runtime/time.go:59 time.Sleep (0x4495b3)
  Goroutine 49 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 50 - User: C:/Go/src/runtime/proc.go:260 runtime.gopark (0x43025f)
  Goroutine 52 - User: e:/go/src/github.com/docker/docker/cli/command/container/hijack.go:101 github.com/docker/docker/cli/command/container.holdHijackedConnection (0x6161c7)
  Goroutine 53 - User: C:/Go/src/runtime/netpoll.go:160 net.runtime_pollWait (0x4291f9)
```

```
(dlv) goroutine 26
Switched from 25 to 26 (thread 976)
(dlv) bt
 0  0x000000000045b667 in runtime.asmcgocall
    at C:/Go/src/runtime/asm_amd64.s:576
 1  0x0000000000405a66 in runtime.cgocall
    at C:/Go/src/runtime/cgocall.go:132
 2  0x00000000004491e6 in syscall.Syscall
    at C:/Go/src/runtime/syscall_windows.go:163
 3  0x0000000000533398 in syscall.CloseHandle
    at C:/Go/src/syscall/zsyscall_windows.go:327
 4  0x000000000048bdec in os.(*file).close
    at C:/Go/src/os/file_windows.go:186
 5  0x000000000048bbee in os.(*File).Close
    at C:/Go/src/os/file_windows.go:168
 6  0x000000000049e3d4 in github.com/docker/docker/cli/command.(*InStream).Close
    at e:/go/src/github.com/docker/docker/cli/command/in.go:26
 7  0x0000000000616ea3 in github.com/docker/docker/cli/command/container.restoreTerminal
    at e:/go/src/github.com/docker/docker/cli/command/container/hijack.go:147
 8  0x0000000000631dea in github.com/docker/docker/cli/command/container.holdHijackedConnection.func2.1
    at e:/go/src/github.com/docker/docker/cli/command/container/hijack.go:56
 9  0x00000000004ebf58 in sync.(*Once).Do
    at C:/Go/src/sync/once.go:44
10  0x0000000000631e3a in github.com/docker/docker/cli/command/container.holdHijackedConnection.func2
    at e:/go/src/github.com/docker/docker/cli/command/container/hijack.go:57
(dlv)
```

### CODE REVIEW

**Comment by darrenstahlmsft** [comment]
Not a maintainer, but LGTM 👍

**Review by vdemeester — APPROVED** [review]
LGTM 🐻
/cc @johnstep @simonferquel

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by johnstep — APPROVED** [review]
LGTM

**Comment by jhowardmsft** [comment]
PPC failures are unrelated I've restarted it a couple of times.

**Comment by thaJeztah** [comment]
Seen this failing on other PRs as well; wondering if it's due to changes at Docker Hub

```
21:42:27 FAIL: docker_cli_push_test.go:607: DockerRegistryAuthTokenSuite.TestPushMisconfiguredTokenServiceResponseError
21:42:27 
21:42:27 docker_cli_push_test.go:615:
21:42:27     c.Assert(out, checker.Contains, "Retrying")
21:42:27 ... obtained string = "" +
21:42:27 ...     "The push refers to a repository [127.0.0.1:5000/busybox]\n" +
21:42:27 ...     "5f70bf18a086: Preparing\n" +
21:42:27 ...     "95f0d78d9341: Preparing\n" +
21:42:27 ...     "toomanyrequests: out of tokens\n"
21:42:27 ... substring string = "Retrying"
21:42:27 
21:42:28 
21:42:28 ----------------------------------------------------------------------
21:42:28 PASS: docker_cli_push_test.go:634: DockerRegistryAuthTokenSuite.TestPushMisconfiguredTokenServiceResponseNoToken	0.172s

**Comment by thaJeztah** [comment]
At a quick glance, it's not using Docker Hub

**Comment by cpuguy83** [comment]
It's failing elsewhere

**Comment by thaJeztah** [comment]
ping @dmcgowan @runcom any ideas?

**Comment by tophj-ibm** [comment]
Opened an issue for the flaky test and am looking into it #32045. Definitely not related to this.

**Comment by thaJeztah** [comment]
yes, lets merge this one
